File: Cpp/Problem_1657.cpp
#include <iostream>
#include <string>
#include <vector>
#include <algorithm>

using namespace std;

// Problem - 1657: https://leetcode.com/problems/determine-if-two-strings-are-close/
// C++ Solution!
class Solution {
public:
    bool closeStrings(string word1, string word2) {
        if (word1.length() != word2.length())
            return false;

        vector<int> freq1(26, 0);
        vector<int> freq2(26, 0);
        for (int i = 0; i < word1.length(); i++) {
            freq1[word1[i] - 'a']++;
            freq2[word2[i] - 'a']++;
        }

        for (int i = 0; i < 26; i++) {
            if (freq1[i] == 0 != freq2[i] == 0)
                return false;
        }

        sort(freq1.begin(), freq1.end());
        sort(freq2.begin(), freq2.end());

        for (int i = 0; i < 26; i++) {
            if (freq1[i] != freq2[i])
                return false;
        }
        return true;
    }
};

int main() {
    Solution sol;

    cout << sol.closeStrings("abc", "bca") << endl;
    cout << sol.closeStrings("a", "aa") << endl;
    cout << sol.closeStrings("cabbba", "abbccc") << endl;

    return 0;
}

File: Python/Problem_1657.py
from collections import Counter

# Problem - 1657: https://leetcode.com/problems/determine-if-two-strings-are-close/
# Python Solution!
class Solution:
    def closeStrings(self, word1: str, word2: str) -> bool:
        if len(word1) != len(word2):
            return False

        freq1 = Counter(word1)
        freq2 = Counter(word2)

        return freq1.keys() == freq2.keys() and sorted(freq1.values()) == sorted(freq2.values())


if __name__=='__main__':
    sol = Solution()

    print(sol.closeStrings("abc", "bca"))
    print(sol.closeStrings("a", "aa"))
    print(sol.closeStrings("cabbba", "abbccc"))